### PR TITLE
ENH: use towncrier to build the release note

### PR DESCRIPTION
**Pull request by mattip**
Add infrastructure and instructions to use towncrier to build the release note from a directory of fragments.

`python -mtowncrier --draft` (once there are some fragments in `./changelog`) will preview the current release note status. 

Running `python -mtowncrier` will overwrite `./doc/release/latest-note.rst`, delete all the fragments from `./changelog`, and stage the changes for a git commit. 

A fragment is a file in `./changelog` with a name like `139999.changes.rst` with a title line and content like

```
Use ``towncrier`` to build the release note
-------------------------------------------
Use ``towncrier`` to build the release note out of fragments in a directory
```

### CODE REVIEW

**Comment by seberg** [comment]
I am wondering if we can do a tiny CI check to avoid bad file naming?

File: doc/RELEASE_WALKTHROUGH.rst.txt [hunk 1]
```diff
@@ -38,6 +38,10 @@ to the maintenance branch, and later will be forward ported to master.
 Finish the Release Note
 -----------------------
 
+.. note:
+
+  This has changed now that we use ``towncrier``
```
Comment (seberg) [inline doc/RELEASE_WALKTHROUGH.rst.txt]: Might be a bit cryptic without information on where to look?

**Review by seberg — APPROVED** [review]
Looks good to me. Lets not worry too much about some extra bells and whistles. It would be good to set this up to make it easier to merge some PRs and other things can be added later. @charris does this seem good to you?

File: changelog/README.rst [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+* ``new_features``: New user facing features like ``kwargs``.
+* ``improvements``: Performance and edge-case changes
+* ``changes``: Other changes
+
+So for example: ``123.c_api.rst``, ``456.new_feature.rst``.
+
+If you are unsure what pull request type to use, don't hesitate to ask in your
+PR.
+
+Note that the ``towncrier`` tool will automatically reflow your text, so it
```
Comment (shoyer) [inline changelog/README.rst]: I don't think this is true for our template, so this should probably be removed  -- you removed the automatic insertion of list items from the pytest template.

File: changelog/README.rst [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+Changelog
+=========
+
+This directory contains "news fragments" which are short files that contain a
+small **ReST**-formatted text that will be added to the next what's new page.
+
+Make sure to use full sentences with correct case and punctuation, and please
+try to use Sphinx intersphinx using backticks. The fragment should have a
+header line and an underline using ``------``
```
Comment (shoyer) [inline changelog/README.rst]: I would be great to add a brief example here, e.g.,

```
my_new_feature option for np.my_favorite_function
-------------------------------------------------
The `my_new_feature` option is now available for `my_favorite_function`.
To use it, write ``np.my_favorite_function(..., my_new_feature=True)``.
```
with a note about how one backtick is used for the names of NumPy functions, and two backticks for code snippets (assuming I have that right).

**Review by shoyer — COMMENTED** [review]
This should definitely make for a better contributor workflow :)

File: changelog/README.rst (hunk 3 above)
Comment (mattip) [reply changelog/README.rst]: adding

File: changelog/README.rst (hunk 2 above)
Comment (mattip) [reply changelog/README.rst]: removing

File: doc/RELEASE_WALKTHROUGH.rst.txt (hunk 1 above)
Comment (mattip) [reply doc/RELEASE_WALKTHROUGH.rst.txt]: adding the path to the `README.rst`

**Comment by mattip** [comment]
It seems towncrier requires a `pyproject.toml` file, but the default `build-system` section causes `pip install .` tries to build numpy without `cython`.

File: setup.py [hunk 4]
```diff
@@ -364,7 +364,7 @@ def parse_setuppy_commands():
 
 
 def setup_package():
-    src_path = os.path.dirname(os.path.abspath(sys.argv[0]))
+    src_path = os.path.dirname(os.path.abspath(__file__))
```
Comment (mattip) [inline setup.py]: When using `pyproject.yaml`, `setup.py` is now called via `exec(compile(code, __file__, 'exec'), locals())` in `python pip/_vendor/pep517/_in_process.py get_requires_for_build_wheel /tmp/tmp6g6mz9bq`. Since it is not called directly on the command line in a subprocess but via `exec`, `sys.argv[0]` is no longer `./setup.py`.

Comment (seberg) [reply setup.py]: yeah, just tracked down this too, but did not manage to quite get to the bottom of everything, maybe we can fix it tomorrow, or push towncrier to give an alternative toml file... (I commented on an issue there)

Comment (eric-wieser) [reply setup.py]: This change looks harmless to me - functions in `setup.py` aren't public API, and any use-case involving symlinks to setup.py seems incredibly contrived

File: pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,60 @@
+[build-system]
+# Minimum requirements for the build system to execute.
+requires = ["setuptools", "wheel", "cython"]  # PEP 508 specification
```
Comment (eric-wieser) [inline pyproject.toml]: It's tempting to do the PEP 508 stuff in a separate PR, just to give it a little more visibility. This looks harmless, but it might be good to seed discussion about any further PEP508 changes

**Comment by mattip** [comment]
Adding the pyproject.toml file seemed non-controversial on the mailing list. Can this go in as-is?

**Comment by seberg** [comment]
I have opened https://github.com/hawkowl/towncrier/pull/156 which adds `--pyproject` to pass in a fixed toml file. That one probably still needs tests so need to figure it out, but it is of course very easy to add to towncrier.

**Comment by mattip** [comment]
When #14053 goes in this will need a rebase

**Comment by seberg** [comment]
OK, rebased and squashed this. Will likely merge soon.

**Comment by seberg** [comment]
OK, I will merge this. happy to revert and clean up everything if someone is unhappy with that (sorry in that case). But there are a few PRs that I would like to get going, and they need release notes, so being unclear about how to write them best right now is not helpful.

**Comment by seberg** [comment]
Ok, I noticed that there is at least one typo and the final release notes are indented a bit funnily (template issue probably). These are small things and I will probably create a quick update PR soon.

**Comment by seberg** [comment]
Hmmm, I am a bit disappointed. It seems to me that towncrier expects release notes to always come in a bullet point format. But we often have longer text and currently use a subheading format. And right now, I think I would prefer to keep it like that. I think I will look into pushing some fixes for that upstream...

Does not need to stop us using it for the moment though, at least we cannot possibly run into issues with merge conflicts...
